README
sysstat: System performance tools for the Linux operating system...
--
(C) 1999-2007 Sebastien Godard (sysstat <at> orange.fr)

The latest version of sysstat can always be found on my web site at:

http://pagesperso-orange.fr/sebastien.godard/

sysstat package is also available at ibiblio's Linux archive
in the following directory:

ftp://ibiblio.org/pub/Linux/system/status/

See the CHANGES file to know the new features/improvements/bug fixes added
in this release of sysstat.


INSTALLATION
------------

The sysstat utilities are a collection of performance monitoring tools for
Linux. These include mpstat, iostat, pidstat, sar, sadc, sadf and sa tools. 

The first stage is to configure sysstat for your system:

	./configure

You can set several variables and parameters on the command line.
Please enter "./configure --help" to display them.
There is another way to configure sysstat instead of entering "./configure":
this is the Interactive Configuration script (iconfig) which will ask you
for the value of the main sysstat variables and parameters.
Enter "./iconfig" then answer the questions or enter Return to accept
the (sane) default values. For yes/no questions, please answer 'y' or 'n'
(without the quotes): It is case sensitive! You can also enter '?' to get
a help message that will explain the meaning of each variable or parameter.

The next stage is to build the various binary files. Enter:

	make

Then log in as root and enter:

	make install

(see next section to know the files that are installed).
That's all!

Of course tell me if there are any problems. This is the only way I can improve
'sysstat'. Please also remember to read the FAQ included in this package.

Patches and suggestions for improvements are always welcome!
Send them to (sysstat <at> orange.fr).


FILES THAT ARE INSTALLED
------------------------

I _hate_ when packages install files everywhere on my disk and I don't know
where...
So here is the list of files installed by sysstat, when you ask for a
complete installation.
${PREFIX} is the value of the PREFIX variable defined in the Makefile
(usually set to /usr/local or /usr).

${PREFIX}/lib/sa/sadc
${PREFIX}/lib/sa/sa1
${PREFIX}/lib/sa/sa2
${PREFIX}/bin/sar
${PREFIX}/bin/sadf
${PREFIX}/bin/iostat
${PREFIX}/bin/mpstat
${PREFIX}/bin/pidstat
${PREFIX}(/share)/man/man8/sadc.8
${PREFIX}(/share)/man/man8/sa1.8
${PREFIX}(/share)/man/man8/sa2.8
${PREFIX}(/share)/man/man1/sar.1
${PREFIX}(/share)/man/man1/sadf.1
${PREFIX}(/share)/man/man1/iostat.1
${PREFIX}(/share)/man/man1/mpstat.1
${PREFIX}(/share)/man/man1/pidstat.1
${PREFIX}/share/locale/*/LC_MESSAGES/sysstat.mo
${PREFIX}/share/doc/sysstat-x.y.z/*
/var/log/sa
${INIT_DIR}/sysstat
/etc/sysconfig/sysstat
/etc/sysconfig/sysstat.ioconf
/etc/cron.d/sysstat
/etc/rc.d/rc.sysstat		for [SLACKWARE]
${RC_DIR}/rc2.d/S03sysstat
${RC_DIR}/rc3.d/S03sysstat
${RC_DIR}/rc5.d/S03sysstat

with:

${INIT_DIR}=/etc/rc.d/init.d/	for [REDHAT] [CALDERA] [MANDRIVA] [TURBOLINUX] [KONDARA]
${INIT_DIR}=/sbin/init.d/	for [SUSE <= 7.0]
${INIT_DIR}=/etc/init.d/	for [SUSE >= 7.1] [DEBIAN]
${RC_DIR}=/etc/rc.d/		for [REDHAT] [CALDERA] [MANDRIVA] [TURBOLINUX] [KONDARA] [SLACKWARE]
${RC_DIR}=/sbin/init.d/		for [SUSE <= 7.0]
${RC_DIR}=/etc/init.d/		for [SUSE >= 7.1]
${RC_DIR}=/etc/			for [DEBIAN]

sysstat may also install some links in ${RC_DIR}/rc[0146].d/ directory
if chkconfig is used.

MISCELLANEOUS
-------------

The sar, mpstat, iostat and pidstat commands are only front-ends to the kernel
proc filesystem... They cannot display statistics that Linux does not
provide, nor can they be more accurate than Linux is.
The sysstat package has been tested on Linux kernels 2.0.x, 2.2.x, 2.4.x and
2.6.x, but since new statistics are added in sysstat, they are not necessarily
present in old kernels.

It has been designed with National Language Support (NLS) in mind, using
the GNU gettext package (available at http://www.gnu.org).
sysstat has been translated into several languages.
Anyway you are welcome if you want to make other translations available ;-)
Please read the README-nls file in the nls directory before.

--
Sebastien Godard (sysstat <at> orange.fr)



CHANGES
Changes:

2010/09/11: Version 9.1.5 - Sebastien Godard (sysstat <at> orange.fr)
	* Added voltage inputs statistics to sar and sadc.
	* Added %temp field to device temperature statistics (sar -m TEMP).
	* Added drpm field to fan speed statistics (sar -m FAN).
	* XSD and DTD documents updated.
	* sar manual page updated. Indicate that svctm field should
	  no longer be trusted.
	* Removed link to isag(1) from man pages.
	* NLS updated. Czech translation added.
	* Sample crontabs and sysstat init script: Don't collect disk
	  data by default.
	* Code cleaned.

2010/07/29: Version 9.1.4 - Sebastien Godard (sysstat <at> orange.fr)
	* [Jan Kaluza]: Added fan speed and device temperature
	  statistics to sar and sadc.
	* [Jan Kaluza]: Configure script updated. Now check for
	  lm_sensors library.
	* Configure script updated: Added --disable-sensors option.
	* Configure script updated: Removed --enable-smp-race option.
	* iconfig script updated.
	* sar manual page updated.
	* XSD and DTD documents updated.
	* [Ivana Varekova]: sysstat init script updated.
	* Default owner for sadc/sar crontab is now root.
	* Various fixes in cifsiostat and nfsiostat manual pages.
	* NLS updated.

2010/06/27: Version 9.1.3 - Sebastien Godard (sysstat <at> orange.fr)
	* [Ivana Varekova]: Move NFS code out from iostat and create
	  the new nfsiostat command.
	* [Ivana Varekova]: Added cifsiostat command. This command
	  displays CIFS statistics.
	* [Mario Konrad]: Added regular expressions support to pidstat's
	  option -C.
	* [Mario Konrad]: Added option -s to pidstat to display stack
	  statistics.
	* [Ivana Varekova]: Fixed condition used by sar to distinguish
	  betwwen a counter reset by a newly registered device and a
	  counter overflow.
	* [Jan Kaluza]: Added --debuginfo option to iostat.
	* [Jan Kaluza]: Added --enable-debuginfo option to configure script.
	* iconfig configuration script updated.
	* iostat manual page updated. Indicate that svctm field should
	  no longer be trusted.
	* pidstat manual page updated.
	* autoconf script updated.
	* NLS updated.
	* README and CREDITS files updated.

2010/05/23: Version 9.1.2 - Sebastien Godard (sysstat <at> orange.fr)
	* Added r_await and w_await fields to iostat's extended statistics.
	* Added support for tickless CPUs in sar and mpstat.
	* NLS was not taken into account when mpstat or sar were displaying
	  some null statistics. This is now fixed.
	* sadc no longer syncs data with its output file contents. It
	  put a pain on flash devices as it undermined any tuning of
	  the vm's write behavior [DEBIAN Bug#559686].
	* NLS updated. Basque translation added.
	* iostat, sar and sa1 manual pages updated.
	* CREDITS file updated.

2010/02/28: Version 9.1.1 - Sebastien Godard (sysstat <at> orange.fr)
	* Remove support for kernels older than 2.6.x.
	* iostat now takes into account POSIXLY_CORRECT environment
	  variable. iostat default output for disk utilization is
	  expressed in kB/s, unless this variable is set (in which case
	  the output is expressed in blocks/s).
	* mpstat can now display per processor software interrupts
	  statistics. This requires Linux kernel 2.6.31 or later.
	* Because of a change in /proc/interrupts file format, mpstat
	  was no longer counting the proper number of interrupts. This
	  is now fixed.
	* Manual pages updated.
	* NLS updated.
	* Code cleaned.

2009/11/11: Version 9.0.6 - Sebastien Godard (sysstat <at> orange.fr)
	* "sadf -x" now takes into account options -s and -e (which
	  specify a starting and ending time) and also interval and
	  count parameters [DEBIAN bug#546259].
	* Option -C added to sadf. Use it to tell sadf to display comments
	  present in file (also works with XML format).
	* sar and sadf sometimes didn't handle COMMENT records properly.
	  This is now fixed.
	* XML output (displayed by sadf -x) modified for kernel tables
	  statistics.
	* XSD and DTD documents updated.
	* [Yibin Shen]: HZ variable was not explicitly set in sadc.c. This
	  is now fixed.
	* [Lans Carstensen]: sargraph added (sargraph is a shell script
	  used to make graphs based on sadf XML output).
	* sadf manual page updated.
	* FAQ updated.
	* NLS updated.
	* CREDITS file updated.

2009/09/20: Version 9.0.5 - Sebastien Godard (sysstat <at> orange.fr)
	* sysstat should now be able to pretty print device names whose
	  minor numbers are greater than or equal to 256. (Tests have
	  been made on a large 26TB RHEL5/PP Linux cluster.)
	* sadc manual page updated.
	* NLS updated.
	* FAQ updated.

2009/07/19: Version 9.0.4 - Sebastien Godard (sysstat <at> orange.fr)
	* [Jonathan Kamens]: Fix double free/memory corruption bug
	  with sadc.
	* [Jeroen Roovers]: Get entirely rid of implicit rule for
	  libsyscom.a in Makefile to fix a problem with parallel
	  execution.
	* sysstat.ioconf configuration file updated.
	* NLS updated.
	* CREDITS file updated.

2009/05/24: Version 9.0.3 - Sebastien Godard (sysstat <at> orange.fr)

sa.h
/*
 * sar/sadc: Report system activity
 * (C) 1999-2009 by Sebastien Godard (sysstat <at> orange.fr)
 */

#ifndef _SA_H
#define _SA_H

#include "common.h"
#include "rd_stats.h"


/*
 ***************************************************************************
 * Activity identification values.
 ***************************************************************************
 */

/* Number of activities */
#define NR_ACT	30

/* Activities */
#define A_CPU		1
#define A_PCSW		2
#define A_IRQ		3
#define A_SWAP		4
#define A_PAGE		5
#define A_IO		6
#define A_MEMORY	7
#define A_KTABLES	8
#define A_QUEUE		9
#define A_SERIAL	10
#define A_DISK		11
#define A_NET_DEV	12
#define A_NET_EDEV	13
#define A_NET_NFS	14
#define A_NET_NFSD	15
#define A_NET_SOCK	16
#define A_NET_IP	17
#define A_NET_EIP	18
#define A_NET_ICMP	19
#define A_NET_EICMP	20
#define A_NET_TCP	21
#define A_NET_ETCP	22
#define A_NET_UDP	23
#define A_NET_SOCK6	24
#define A_NET_IP6	25
#define A_NET_EIP6	26
#define A_NET_ICMP6	27
#define A_NET_EICMP6	28
#define A_NET_UDP6	29
#define A_PWR_CPUFREQ	30


/* Macro used to flag an activity that should be collected */
#define COLLECT_ACTIVITY(m)	act[get_activity_position(act, m)]->options |= AO_COLLECTED

/* Macro used to flag an activity that should be selected */
#define SELECT_ACTIVITY(m)	act[get_activity_position(act, m)]->options |= AO_SELECTED


/*
 ***************************************************************************
 * Flags.
 ***************************************************************************
 */

#define S_F_SINCE_BOOT		0x00000001
#define S_F_SA_ROTAT      	0x00000002
#define S_F_DEV_PRETTY		0x00000004
#define S_F_FORCE_FILE		0x00000008
#define S_F_INTERVAL_SET	0x00000010
#define S_F_TRUE_TIME		0x00000020
#define S_F_LOCK_FILE		0x00000040
#define S_F_HAS_PPARTITIONS	0x00000080
#define S_F_HAS_DISKSTATS	0x00000100
#define S_F_FILE_LOCKED		0x00000200
#define S_F_PER_PROC		0x00000400
#define S_F_HORIZONTALLY	0x00000800
#define S_F_COMMENT		0x00001000

#define WANT_SINCE_BOOT(m)	(((m) & S_F_SINCE_BOOT)      == S_F_SINCE_BOOT)
#define WANT_SA_ROTAT(m)	(((m) & S_F_SA_ROTAT)        == S_F_SA_ROTAT)
#define USE_PRETTY_OPTION(m)	(((m) & S_F_DEV_PRETTY)      == S_F_DEV_PRETTY)
#define FORCE_FILE(m)		(((m) & S_F_FORCE_FILE)      == S_F_FORCE_FILE)
#define INTERVAL_SET(m)		(((m) & S_F_INTERVAL_SET)    == S_F_INTERVAL_SET)
#define PRINT_TRUE_TIME(m)	(((m) & S_F_TRUE_TIME)       == S_F_TRUE_TIME)
#define LOCK_FILE(m)		(((m) & S_F_LOCK_FILE)       == S_F_LOCK_FILE)
#define HAS_PPARTITIONS(m)	(((m) & S_F_HAS_PPARTITIONS) == S_F_HAS_PPARTITIONS)
#define HAS_DISKSTATS(m)	(((m) & S_F_HAS_DISKSTATS)   == S_F_HAS_DISKSTATS)
#define FILE_LOCKED(m)		(((m) & S_F_FILE_LOCKED)     == S_F_FILE_LOCKED)
#define WANT_PER_PROC(m)	(((m) & S_F_PER_PROC)        == S_F_PER_PROC)
#define DISPLAY_HORIZONTALLY(m)	(((m) & S_F_HORIZONTALLY)    == S_F_HORIZONTALLY)
#define DISPLAY_COMMENT(m)	(((m) & S_F_COMMENT)         == S_F_COMMENT)

/* Output flags for options -R / -r / -S */
#define AO_F_MEM_DIA		0x00000001
#define AO_F_MEM_AMT		0x00000002
#define AO_F_MEM_SWAP		0x00000004

#define DISPLAY_MEMORY(m)	(((m) & AO_F_MEM_DIA)        == AO_F_MEM_DIA)
#define DISPLAY_MEM_AMT(m)	(((m) & AO_F_MEM_AMT)        == AO_F_MEM_AMT)
#define DISPLAY_SWAP(m)		(((m) & AO_F_MEM_SWAP)       == AO_F_MEM_SWAP)

/* Output flags for option -u [ ALL ] */
#define AO_F_CPU_DEF		0x00000001
#define AO_F_CPU_ALL		0x00000002

#define DISPLAY_CPU_DEF(m)	(((m) & AO_F_CPU_DEF)        == AO_F_CPU_DEF)
#define DISPLAY_CPU_ALL(m)	(((m) & AO_F_CPU_ALL)        == AO_F_CPU_ALL)

/* Output flags for option -d */
#define AO_F_DISK_PART		0x00000001

#define COLLECT_PARTITIONS(m)	(((m) & AO_F_DISK_PART)      == AO_F_DISK_PART)

/*
 ***************************************************************************
 * Various keywords and constants.
 ***************************************************************************
 */

/* Keywords */
#define K_XALL		"XALL"
#define K_SUM		"SUM"
#define K_DEV		"DEV"
#define K_EDEV		"EDEV"
#define K_NFS		"NFS"
#define K_NFSD		"NFSD"
#define K_SOCK		"SOCK"
#define K_IP		"IP"
#define K_EIP		"EIP"
#define K_ICMP		"ICMP"
#define K_EICMP		"EICMP"
#define K_TCP		"TCP"
#define K_ETCP		"ETCP"
#define K_UDP		"UDP"
#define K_DISK		"DISK"
#define K_INT		"INT"
#define K_SNMP		"SNMP"
#define K_SOCK6		"SOCK6"
#define K_IP6		"IP6"
#define K_EIP6		"EIP6"
#define K_ICMP6		"ICMP6"
#define K_EICMP6	"EICMP6"
#define K_UDP6		"UDP6"
#define K_IPV6		"IPV6"
#define K_POWER		"POWER"
#define K_XDISK		"XDISK"

/* sadc program */
#define SADC		"sadc"

/* Time must have the format HH:MM:SS with HH in 24-hour format */
#define DEF_TMSTART	"08:00:00"
#define DEF_TMEND	"18:00:00"

/*
 * Macro used to define activity bitmap size.
 * All those bitmaps have an additional bit used for global activity
 * (eg. CPU "all" or total number of interrupts). That's why we do "(m) + 1".
 */
#define BITMAP_SIZE(m)	((((m) + 1) / 8) + 1)

/* Pre-allocation constants */
#define NR_IFACE_PREALLOC	2
#define NR_SERIAL_PREALLOC	2
#define NR_DISK_PREALLOC	3

#define UTSNAME_LEN		65
#define TIMESTAMP_LEN		16
#define XML_TIMESTAMP_LEN	64
#define HEADER_LINE_LEN		512

/* Maximum number of args that can be passed to sadc */
#define MAX_ARGV_NR	32

/* Miscellaneous constants */
#define USE_SADC	0
#define USE_SA_FILE	1
#define NO_TM_START	0
#define NO_TM_END	0
#define NO_RESET	0
#define NON_FATAL	0
#define FATAL		1
#define C_SAR		0
#define C_SADF		1
#define ALL_ACTIVITIES	~0U

#define SOFT_SIZE	0
#define HARD_SIZE	1

#define CLOSE_XML_MARKUP	0
#define OPEN_XML_MARKUP		1

#define COUNT_ACTIVITIES	0
#define COUNT_OUTPUTS		1


/*
 ***************************************************************************
 * Generic description of an activity.
 ***************************************************************************
 */

/* Activity options */
#define AO_NULL			0x00
/*
 * Indicate that corresponding activity should be collected by sadc.
 */
#define AO_COLLECTED		0x01
/*
 * Indicate that corresponding activity should be displayed by sar.
 */
#define AO_SELECTED		0x02
/*
 * Indicate that, when registered again, activity counters will get back
 * the values they had when they were unregistered (eg. CPUs, which can
 * be disabled/enabled on the fly).
 */
#define AO_REMANENT		0x04
/*
 * Indicate that the interval of time, given to f_print() function
 * displaying statistics, should be the interval of time in jiffies
 * multiplied by the number of processors.
 */
#define AO_GLOBAL_ITV		0x08
/*
 * This flag should be set for every activity closing a markup used
 * by several activities. Used by sadf f_xml_print() functions to
 * display XML output.
 */
#define AO_CLOSE_MARKUP		0x10
/*
 * Indicate that corresponding activity has multiple different
 * output formats. This is the case for example for memory activity
 * with options -r and -R.
 */
#define AO_MULTIPLE_OUTPUTS	0x20

#define IS_COLLECTED(m)		(((m) & AO_COLLECTED)        == AO_COLLECTED)
#define IS_SELECTED(m)		(((m) & AO_SELECTED)         == AO_SELECTED)
#define IS_REMANENT(m)		(((m) & AO_REMANENT)         == AO_REMANENT)
#define NEEDS_GLOBAL_ITV(m)	(((m) & AO_GLOBAL_ITV)       == AO_GLOBAL_ITV)
#define CLOSE_MARKUP(m)		(((m) & AO_CLOSE_MARKUP)     == AO_CLOSE_MARKUP)
#define HAS_MULTIPLE_OUTPUTS(m)	(((m) & AO_MULTIPLE_OUTPUTS) == AO_MULTIPLE_OUTPUTS)


/* Type for all functions counting items */
#define __nr_t		int
/* Type for all functions reading statistics */
#define __read_funct_t	void
/* Type for all functions displaying statistics */
#define __print_funct_t void

#define _buf0	buf[0]

/* Structure used to define a bitmap needed by an activity */
struct act_bitmap {
	/*
	 * Bitmap for activities that need one. Remember to allocate it
	 * before use!
	 */
	unsigned char *b_array;
	/*
	 * Size of the bitmap in bits. In fact, bitmap is sized to bitmap_size + 1
	 * to take into account CPU "all"
	 */
	int b_size;
};

/* Structure used to define an activity */
struct activity {
	/*
	 * This variable contains the identification value (A_...) for this activity.
	 */
	unsigned int id;
	/*
	 * Activity options (AO_SELECTED, ...)
	 */
	unsigned int options;
	/*
	 * The f_count() function is used to count the number of
	 * items (serial lines, network interfaces, etc.).
	 * Such a function should _always_ return a value greater than
	 * or equal to 0.
	 *
	 * A NULL value for this function pointer indicates that the number of items
	 * is a constant (and @nr is set to this value).
	 *
	 * This function is called even if activity has not been selected, to make
	 * sure that all items have been calculated (including #CPU, etc.)
	 */
	__nr_t (*f_count) (struct activity *);
	/*
	 * This function reads the relevant file and fill the buffer
	 * with statistics corresponding to given activity.
	 */
	__read_funct_t (*f_read) (struct activity *);
	/*
	 * This function displays activity statistics onto the screen.
	 */
	__print_funct_t (*f_print) (struct activity *, int, int, unsigned long long);
	/*
	 * This function displays average activity statistics onto the screen.
	 */
	__print_funct_t (*f_print_avg) (struct activity *, int, int, unsigned long long);
	/*
	 * This function is used by sadf to display activity in a format that can
	 * easily be ingested by a relational database, or a format that can be
	 * handled by pattern processing commands like "awk".
	 */
	__print_funct_t (*f_render) (struct activity *, int, char *, int, unsigned long long);
	/*
	 * This function is used by sadf to display activity statistics in XML.
	 */
	__print_funct_t (*f_xml_print) (struct activity *, int, int, unsigned long long);
	/*
	 * Header string displayed by sadf -d/-D.
	 */
	char *hdr_line;
	/*
	 * Name of activity.
	 */
	char *name;
	/*
	 * Number of item.
	 */
	__nr_t *nr;
	/*
	 * Size of an item.
	 * This is the size of the corresponding structure, as read from or written
	 * to a file, or read from or written by the data collector.
	 */
	int fsize;
	/*
	 * Size of an item.
	 * This is the size of the corresponding structure as mapped into memory.
	 * @msize can be different from @fsize when data are read from or written to
	 * a data file from a different sysstat version.
	 */
	int msize;
	/*
	 * Optional flags for activity. This is eg. used when AO_MULTIPLE_OUTPUTS
	 * option is set.
	 */
	unsigned int opt_flags;
	/*
	 * Buffers that will contain the statistics read. Its size is @nr * @size each.
	 * [0]: used by sadc. Used by sar to save first collected stats (used later to
	 * compute average).
	 * [1] and [2]: current/previous statistics values (used by sar).
	 */
	void *buf[3];
	/*
	 * Bitmap for activities that need one. Such a bitmap is needed by activity
	 * if @bitmap is not NULL.
	 */
	struct act_bitmap *bitmap;
};


/*
 ***************************************************************************
 * Definitions of header structures.
 *
 * Format of system activity data files:
 *	 __
 *	|
 *	| file_magic structure
 *	|
 * 	|--
 *	|
 * 	| file_header structure
 * 	|
 * 	|--                         --|
 * 	|                             |
 * 	| file_activity structure     | * sa_nr_act
 * 	|                             |
 * 	|--                         --|
 * 	|                             |
 * 	| record_header structure     |
 * 	|                             |
 * 	|--                           | * <count>
 * 	|                             |
 * 	| Statistics structures...(*) |
 * 	|                             |
 * 	|--                         --|
 *
 * (*)Note: If it's a special record, we may find a comment instead of
 * statistics (R_COMMENT record type) or even nothing at all (R_RESTART
 * record type).
 ***************************************************************************
 */

/*
 * Sysstat magic number. Should never be modified.
 * Indicate that the file was created by sysstat.
 */
#define SYSSTAT_MAGIC	0xd596

/*
 * Datafile format magic number.
 * Modified to indicate that the format of the file is
 * no longer compatible with that of previous sysstat versions.
 */
#define FORMAT_MAGIC	0x2170

/* Structure for file magic header data */
struct file_magic {
	/*
	 * This field identifies the file as a file created by sysstat.
	 */
	unsigned short sysstat_magic;
	/*
	 * The value of this field varies whenever datafile format changes.
	 */
	unsigned short format_magic;
	/*
	 * Sysstat version used to create the file.
	 */
	unsigned char  sysstat_version;
	unsigned char  sysstat_patchlevel;
	unsigned char  sysstat_sublevel;
	unsigned char  sysstat_extraversion;
};

#define FILE_MAGIC_SIZE	(sizeof(struct file_magic))


/* Header structure for system activity data file */
struct file_header {
	/*
	 * Timestamp in seconds since the epoch.
	 */
	unsigned long sa_ust_time	__attribute__ ((aligned (8)));
	/*
	 * Number of activities saved in the file
	 */
	unsigned int sa_nr_act		__attribute__ ((aligned (8)));
	/*
	 * Current day, month and year.
	 * No need to save DST (daylight saving time) flag, since it is not taken
	 * into account by the strftime() function used to print the timestamp.
	 */
	unsigned char sa_day;
	unsigned char sa_month;
	unsigned char sa_year;
	/*
	 * Size of a long integer. Useful to know the architecture on which
	 * the datafile was created.
	 */
	char sa_sizeof_long;
	/*
	 * Operating system name.
	 */
	char sa_sysname[UTSNAME_LEN];
	/*
	 * Machine hostname.
	 */
	char sa_nodename[UTSNAME_LEN];
	/*
	 * Operating system release number.
	 */
	char sa_release[UTSNAME_LEN];
	/*
	 * Machine architecture.
	 */
	char sa_machine[UTSNAME_LEN];
};

#define FILE_HEADER_SIZE	(sizeof(struct file_header))


/* List of activities saved in file */
struct file_activity {
	/*
	 * Identification value of activity.
	 */
	unsigned int id		__attribute__ ((aligned (4)));
	/*
	 * Number of items for this activity.
	 */
	__nr_t nr		__attribute__ ((packed));
	/*
	 * Size of an item structure.
	 */
	int size		__attribute__ ((packed));
};

#define FILE_ACTIVITY_SIZE	(sizeof(struct file_activity))


/* Record type */
/*
 * R_STATS means that this is a record of statistics.
 */
#define R_STATS		1
/*
 * R_RESTART means that this is a special record containing
 * a LINUX RESTART message.
 */
#define R_RESTART	2
/*
 * R_LAST_STATS warns sar that this is the last record to be written
 * to file before a file rotation, and that the next data to come will
 * be a header file.
 * Such a record is tagged R_STATS anyway before being written to file.
 */
#define R_LAST_STATS	3
/*
 * R_COMMENT means that this is a special record containing
 * a comment.
 */
#define R_COMMENT	4

/* Maximum length of a comment */
#define MAX_COMMENT_LEN	64

/* Header structure for every record */
struct record_header {
	/*
	 * Machine uptime (multiplied by the # of proc).
	 */
	unsigned long long uptime	__attribute__ ((aligned (16)));
	/*
	 * Uptime reduced to one processor. Always set, even on UP machines.
	 */
	unsigned long long uptime0	__attribute__ ((aligned (16)));
	/*
	 * Timestamp (number of seconds since the epoch).
	 */
	unsigned long ust_time		__attribute__ ((aligned (16)));
	/*
	 * Record type: R_STATS, R_RESTART,...
	 */
	unsigned char record_type	__attribute__ ((aligned (8)));
	/*
	 * Timestamp: Hour (0-23), minute (0-59) and second (0-59).
	 * Used to determine TRUE time (immutable, non locale dependent time).
	 */
	unsigned char hour;
	unsigned char minute;
	unsigned char second;
};

#define RECORD_HEADER_SIZE	(sizeof(struct record_header))


/*
 ***************************************************************************
 * Macro functions definitions.
 *
 * Note: Using 'do ... while' makes the macros safer to use
 * (remember that macro use are followed by a semicolon).
 ***************************************************************************
 */

/* Close file descriptors */
#define CLOSE_ALL(_fd_)		do {			\
					close(_fd_[0]); \
					close(_fd_[1]); \
				} while (0)

#define CLOSE(_fd_)		if (_fd_ >= 0)		\
					close(_fd_)


/*
 ***************************************************************************
 * Various structure definitions.
 ***************************************************************************
 */

/* Structure for timestamps */
struct tstamp {
	int tm_sec;
	int tm_min;
	int tm_hour;
	int use;
};


/*
 ***************************************************************************
 * Functions prototypes.
 ***************************************************************************
 */

/* Functions used to count number of items */
extern __nr_t
	wrap_get_cpu_nr(struct activity *);
extern __nr_t
	wrap_get_irq_nr(struct activity *);
extern __nr_t
	wrap_get_serial_nr(struct activity *);
extern __nr_t
	wrap_get_disk_nr(struct activity *);
extern __nr_t
	wrap_get_iface_nr(struct activity *);

/* Functions used to read activities statistics */
extern __read_funct_t
	wrap_read_stat_cpu(struct activity *);
extern __read_funct_t
	wrap_read_stat_pcsw(struct activity *);
extern __read_funct_t
	wrap_read_stat_irq(struct activity *);
extern __read_funct_t
	wrap_read_swap(struct activity *);
extern __read_funct_t
	wrap_read_paging(struct activity *);
extern __read_funct_t
	wrap_read_io(struct activity *);
extern __read_funct_t
	wrap_read_meminfo(struct activity *);
extern __read_funct_t
	wrap_read_kernel_tables(struct activity *);
extern __read_funct_t
	wrap_read_loadavg(struct activity *);
extern __read_funct_t
	wrap_read_tty_driver_serial(struct activity *);
extern __read_funct_t
	wrap_read_disk(struct activity *);
extern __read_funct_t
	wrap_read_net_dev(struct activity *);
extern __read_funct_t
	wrap_read_net_edev(struct activity *);
extern __read_funct_t
	wrap_read_net_nfs(struct activity *);
extern __read_funct_t
	wrap_read_net_nfsd(struct activity *);
extern __read_funct_t
	wrap_read_net_sock(struct activity *);
extern __read_funct_t
	wrap_read_net_ip(struct activity *);
extern __read_funct_t
	wrap_read_net_eip(struct activity *);
extern __read_funct_t
	wrap_read_net_icmp(struct activity *);
extern __read_funct_t
	wrap_read_net_eicmp(struct activity *);
extern __read_funct_t
	wrap_read_net_tcp(struct activity *);
extern __read_funct_t
	wrap_read_net_etcp(struct activity *);
extern __read_funct_t
	wrap_read_net_udp(struct activity *);
extern __read_funct_t
	wrap_read_net_sock6(struct activity *);
extern __read_funct_t
	wrap_read_net_ip6(struct activity *);
extern __read_funct_t
	wrap_read_net_eip6(struct activity *);
extern __read_funct_t
	wrap_read_net_icmp6(struct activity *);
extern __read_funct_t
	wrap_read_net_eicmp6(struct activity *);
extern __read_funct_t
	wrap_read_net_udp6(struct activity *);
extern __read_funct_t
	wrap_read_cpuinfo(struct activity *);

/* Other functions */
extern void
	allocate_bitmaps(struct activity * []);
extern void
	allocate_structures(struct activity * []);
extern int
	check_disk_reg(struct activity *, int, int, int);
extern void
	check_file_actlst(int *, char *, struct activity * [], struct file_magic *,
			  struct file_header *, struct file_activity **,
			  unsigned int [], int);
extern unsigned int
	check_net_dev_reg(struct activity *, int, int, unsigned int);
extern unsigned int
	check_net_edev_reg(struct activity *, int, int, unsigned int);
extern void
	copy_structures(struct activity * [], unsigned int [],
			struct record_header [], int, int);
extern int
	datecmp(struct tm *, struct tstamp *);
extern void
	display_sa_file_version(struct file_magic *);
extern void
	free_bitmaps(struct activity * []);
extern void
	free_structures(struct activity * []);
extern int
	get_activity_nr(struct activity * [], unsigned int, int);
extern int
	get_activity_position(struct activity * [], unsigned int);
extern char *
	get_devname(unsigned int, unsigned int, int);
extern void
	get_itv_value(struct record_header *, struct record_header *,
		      unsigned int, unsigned long long *, unsigned long long *);
extern void
	handle_invalid_sa_file(int *, struct file_magic *, char *, int);
extern int
	next_slice(unsigned long long, unsigned long long, int, long);
extern int
	parse_sar_opt(char * [], int *, struct activity * [], unsigned int *, int);
extern int
	parse_sar_I_opt(char * [], int *, struct activity * []);
extern int
	parse_sa_P_opt(char * [], int *, unsigned int *, struct activity * []);
extern int
	parse_sar_n_opt(char * [], int *, struct activity * []);
extern int
	parse_timestamp(char * [], int *, struct tstamp *, const char *);
extern void
	print_report_hdr(unsigned int, struct tm *, struct file_header *, int);
extern void
	read_file_stat_bunch(struct activity * [], int, int, int, struct file_activity *);
extern int
	sa_fread(int, void *, int, int);
extern void
	select_all_activities(struct activity * []);
extern void
	select_default_activity(struct activity * []);
extern void
	set_bitmap(unsigned char [], unsigned char, unsigned int);
extern void
	set_default_file(struct tm *, char *);
extern void
	set_hdr_rectime(unsigned int, struct tm *, struct file_header *);

#endif  /* _SA_H */
